core/src/databases/database.rs

```diff
@@ -16,9 +40,106 @@ impl Database {
+        match self.db_type {
+            DatabaseType::REMOTE => Err(anyhow!("Remote DB not implemented.")),
+            DatabaseType::LOCAL => {
+                let (tables, _) = store
+                    .list_databases_tables(&project, &self.data_source_id, &self.database_id, None)
+                    .await?;
+
+                let mut table_rows_futures = Vec::new();
+
+                for table in &tables {
+                    let store_ref = &store;
+                    let project_ref = &project;
+                    let data_source_id = &self.data_source_id;
+                    let database_id = &self.database_id;
+                    let table_id = table.table_id().to_string();
+
+                    let rows_future = async move {
+                        store_ref
+                            .list_database_rows(
+                                project_ref,
+                                data_source_id,
+                                database_id,
+                                &table_id,
+                                None,
+                            )
+                            .await
+                    };
+
+                    table_rows_futures.push(rows_future);
+                }
+
+                // Now, we concurrently wait for all futures to complete.
+                let results: Vec<_> = futures::future::join_all(table_rows_futures).await;
+
+                let mut table_rows = HashMap::new();
+
+                for result in results {
+                    match result {
+                        Ok((rows, _)) => {
+                            let first_row = rows.first();
+                            if let Some(row) = first_row {
+                                table_rows.insert(row.table_id().to_string(), rows);
+                            }
+                        }
+                        Err(e) => return Err(e.into()),
+                    }
+                }
+
+                let table_by_id = tables
+                    .iter()
+                    .map(|table| (table.table_id().to_string(), table))
+                    .collect::<HashMap<String, &DatabaseTable>>();
+
+                let table_ids_with_rows = table_rows.keys().collect::<Vec<&String>>();
+
+                let schema_tables = table_ids_with_rows
+                    .par_iter()
+                    .map(|table_id| {
+                        let table = *table_by_id.get(*table_id).unwrap();
+                        let rows = table_rows.get(*table_id).unwrap();
+
+                        let row_contents =
+                            (rows.iter().map(|x| x.content())).collect::<Vec<&Value>>();
+                        let table_schema = TableSchema::from_rows(&row_contents)?;
+                        Ok(DatabaseSchemaTable::new(table.clone(), table_schema))
+                    })
+                    .collect::<Result<Vec<DatabaseSchemaTable>>>()?;
+
+                let mut schema = HashMap::new();
```

Comment:
You can do that by return a tuple in the function above and a properly typed collect at the end